### PR TITLE
Add HMAC One Time Password TA and host application

### PR DESCRIPTION
**Pull request by jbech-linaro**
This implements a HOTP according to RFC4226. The added README.md in the hotp folder will tell you about the details.

This also serves as a good example of how to actually do a plain HMAC using the GP Internal Core API.

### CODE REVIEW

File: hotp/ta/hotp_ta.c [hunk 1]
```diff
@@ -0,0 +1,277 @@
...
+	 ((uint64_t)(v) >> 56))
+
+/* The size of a SHA1 hash in bytes. */
+#define SHA1_HASH_SIZE 20
+
+/* GP says that for HMAC SHA-1, max is 512 bits and min 80 bits. */
+#define MAX_KEY_SIZE 64 /* In bytes */
+#define MIN_KEY_SIZE 10 /* In bytes */
+
+/* Dynamic Binary Code 2 Modulu, which is 10^6 according to the spec. */
```
Comment (jbech-linaro) [inline hotp/ta/hotp_ta.c]: s/Modulu/Modulo/

File: hotp/ta/hotp_ta.c [hunk 2]
```diff
@@ -0,0 +1,277 @@
+/*
+ * Copyright (c) 2017, Linaro Limited
+ * All rights reserved.
+ *
+ * Redistribution and use in source and binary forms, with or without
```
Comment (jenswi-linaro) [inline hotp/ta/hotp_ta.c]: Replacing this with
```
SPDX-License-Identifier: BSD-2-Clause
```
would save a few lines.

File: hotp/ta/hotp_ta.c [hunk 3]
```diff
@@ -0,0 +1,277 @@
...
+		void __maybe_unused **sess_ctx)
+{
+	uint32_t exp_param_types = TEE_PARAM_TYPES(TEE_PARAM_TYPE_NONE,
+						   TEE_PARAM_TYPE_NONE,
+						   TEE_PARAM_TYPE_NONE,
+						   TEE_PARAM_TYPE_NONE);
+	if (param_types != exp_param_types)
+		return TEE_ERROR_BAD_PARAMETERS;
+
+	(void)params;
```
Comment (jenswi-linaro) [inline hotp/ta/hotp_ta.c]: `params` and `sess_ctx` are already tagged as `__maybe_unused`, why is this needed too?

File: hotp/ta/hotp_ta.c [hunk 4]
```diff
@@ -0,0 +1,277 @@
...
+ * The HOTP counter should be a 8 byte array where byte 0 is the MSB and byte 7
+ * is the LSB. To make it easier to update the variable, just simply use a
+ * uint64_t instead. Since the HOTP counter needs to be in big endian, we need
+ * to swap it when incrementing.
+ */
+static uint64_t counter_be = 0;
+
+static void inc_counter(uint64_t *c_be)
+{
+	uint64_t c_le = SWAP64(*c_be);
```
Comment (jenswi-linaro) [inline hotp/ta/hotp_ta.c]: Byteswapping is fun, but I'd rather see a conversion from 64-bit Big Endian to native endian.

File: hotp/ta/hotp_ta.c (hunk 3 above)
Comment (jforissier) [reply hotp/ta/hotp_ta.c]: +1 and tag should be `__unused`

File: hotp/ta/hotp_ta.c (hunk 4 above)
Comment (jforissier) [reply hotp/ta/hotp_ta.c]: ;-)

File: hotp/ta/hotp_ta.c [hunk 5]
```diff
@@ -0,0 +1,277 @@
...
+ * This macro checks the expected parameters, if they are not the same, then it
+ * will set 'res' to TEE_ERROR_BAD_PARAMETERS and will goto label 'label'.
+ */
+#define CHECK_EXP_PARAM_EXIT(label, res, pt, ept) \
+	if (pt != ept) { \
+		EMSG("Expected: 0x%x, got: 0x%x", ept, pt); \
+		res = TEE_ERROR_BAD_PARAMETERS; \
+		goto label; }
+
+#define SWAP64(v) \
```
Comment (jforissier) [inline hotp/ta/hotp_ta.c]: We already have such a macro in a header accessible to TAs, if I'm not mistaken.

File: hotp/ta/hotp_ta.c (hunk 2 above)
Comment (jbech-linaro) [reply hotp/ta/hotp_ta.c]: Have we adopted SPDX now? Even though I know LinuxFoundation (Kate) has been asking for that quite a while and I don't mind, it'd be good to have a formal decision saying that this should be added to all new files (I can only see it in two files in the entire OP-TEE project).

File: hotp/ta/hotp_ta.c (hunk 4 above)
Comment (jbech-linaro) [reply hotp/ta/hotp_ta.c]: Are you looking for something like `__be64_to_cpup` in Linux kernel? If not I don't follow. Anyhow, I've gone back to how I did it in earlier commits, a bit more magic, but less code any no issues with endianness.

File: hotp/ta/hotp_ta.c (hunk 3 above)
Comment (jbech-linaro) [reply hotp/ta/hotp_ta.c]: > @jenswi-linaro why is this needed too?

Copy/paste from the other TA's in this git, but I agree.

> @jforissier +1 and tag should be `__unused`

Isn't `__maybe_unused` better in this case, since it is likely that one would use them in the future. If not we can get rid of / stop using that attribute, since you will **always** know if it will be used or not.

File: hotp/ta/hotp_ta.c (hunk 5 above)
Comment (jbech-linaro) [reply hotp/ta/hotp_ta.c]: I think I tried using `TEE_U64_BSWAP`, but that didn't work I think (only visible to utee?). Anyhow, it's not needed / used anymore since I went back to my old way of doing the increment.

File: hotp/ta/hotp_ta.c (hunk 2 above)
Comment (jforissier) [reply hotp/ta/hotp_ta.c]: Yep let's use SPDX whenever we can. There are 15 files with a SPDX tag in optee_os master already.

File: hotp/ta/hotp_ta.c [hunk 6]
```diff
@@ -0,0 +1,234 @@
...
+
+/* Dynamic Binary Code 2 Modulo, which is 10^6 according to the spec. */
+#define DBC2_MODULO 1000000
+
+/*
+ * Currently this only supports a single key, in the future this could be
+ * updated to support multiple users, all with different unique keys (stored
+ * using secure storage).
+ */
+static uint8_t K[MAX_KEY_SIZE];
```
Comment (jenswi-linaro) [inline hotp/ta/hotp_ta.c]: Not that it matters much, but we're usually using only lower case letters in variables.

File: hotp/ta/hotp_ta.c [hunk 7]
```diff
@@ -0,0 +1,234 @@
...
+static uint32_t K_len;
+
+/* The counter as defined by RFC4226. */
+static uint8_t counter[] = { 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0 };
```
Comment (jenswi-linaro) [inline hotp/ta/hotp_ta.c]: It's a matter of taste, but I would have written this as: `static uint8_t counter[8];`

File: hotp/ta/hotp_ta.c [hunk 8]
```diff
@@ -0,0 +1,234 @@
...
+
+	CHECK_EXP_PARAM_EXIT(exit, res, param_types, exp_param_types);
+
+	CHECK_EXIT(exit, res,
+		   hmac_sha1(K, K_len, counter, sizeof(counter), mac,
+			     &mac_len));
+
+	/* Increment the counter. */
+	for (i = sizeof(counter) - 1; i >= 0; i--) {
+		if (++counter[i] & 0xff)
```
Comment (jenswi-linaro) [inline hotp/ta/hotp_ta.c]: `counter[i]` is an `uint8_t` so `& 0xff` changes nothing.

File: hotp/ta/hotp_ta.c [hunk 9]
```diff
@@ -0,0 +1,234 @@
...
+{
+	return TEE_SUCCESS;
+}
+
+void TA_DestroyEntryPoint(void)
+{
+}
+
+TEE_Result TA_OpenSessionEntryPoint(uint32_t param_types,
+		TEE_Param __maybe_unused params[4],
```
Comment (jenswi-linaro) [inline hotp/ta/hotp_ta.c]: Please align the parameters
Since `params` isn't used at all use `__unused` instead.

File: hotp/ta/hotp_ta.c (hunk 6 above)
Comment (jbech-linaro) [reply hotp/ta/hotp_ta.c]: Yes, but since this is an example I wanted to keep the variables as defined in the spec (and I think I've been almost consistent doing that besides the counter).

File: hotp/host/main.c [hunk 10]
```diff
@@ -0,0 +1,116 @@
...
+};
+
+int main(int argc, char *argv[])
+{
+	TEEC_Context ctx;
+	TEEC_Operation op = { 0 };
+	TEEC_Result res;
+	TEEC_Session sess;
+	TEEC_UUID uuid = TA_HOTP_UUID;
+
```
Comment (jenswi-linaro) [inline hotp/host/main.c]: Empty line not needed

File: hotp/host/main.c [hunk 11]
```diff
@@ -0,0 +1,116 @@
...
+	int i = 0;
```
Comment (jenswi-linaro) [inline hotp/host/main.c]: No need to initialize `i` here

File: hotp/host/main.c [hunk 12]
```diff
@@ -0,0 +1,116 @@
...
+	uint32_t err_origin;
+	uint32_t hotp_value = 0;
```
Comment (jenswi-linaro) [inline hotp/host/main.c]: No need to initialize `hotp_value` here

File: hotp/ta/hotp_ta.c [hunk 13]
```diff
@@ -0,0 +1,234 @@
...
+#include <string.h>
+#include <tee_internal_api_extensions.h>
+#include <tee_internal_api.h>
+
+/*
+ * This macro checks whether the return value is not equal to TEE_SUCCESS, if it
+ * is not equal, then it will set the return value to 'res' and then goto
+ * 'label'.
+ */
+#define CHECK_EXIT(label, res, ret_val) \
```
Comment (jenswi-linaro) [inline hotp/ta/hotp_ta.c]: I have my doubts about these two macros. Personally I don't think it improves readability, and also when reading the code it's not so clear what's happening.

Then there's also the issue with `ret_val` being evaluated three times, doesn't seem proper since it's really a function call with side effects.

File: hotp/ta/hotp_ta.c [hunk 14]
```diff
@@ -0,0 +1,234 @@
...
+		(hmac_result[offset+2] & 0xff) <<  8 |
+		(hmac_result[offset+3] & 0xff);
+
+	*bin_code %= DBC2_MODULO;
+}
+
+static TEE_Result register_shared_key(uint32_t param_types, TEE_Param params[4])
+{
+	TEE_Result res = TEE_SUCCESS;
+
```
Comment (jenswi-linaro) [inline hotp/ta/hotp_ta.c]: Empty line not needed

File: hotp/ta/hotp_ta.c [hunk 15]
```diff
@@ -0,0 +1,234 @@
...
+	return res;
+}
+
+static TEE_Result get_hotp(uint32_t param_types, TEE_Param params[4])
+{
+	TEE_Result res = TEE_SUCCESS;
+	uint32_t hotp_val = 0;
+	uint8_t mac[SHA1_HASH_SIZE];
+	size_t mac_len = sizeof(mac);
+	int i = 0;
```
Comment (jenswi-linaro) [inline hotp/ta/hotp_ta.c]: Initialization not needed.

File: hotp/ta/hotp_ta.c [hunk 16]
```diff
@@ -0,0 +1,234 @@
...
+	K_len = params[0].memref.size;
+	DMSG("Got shared key %s (%u bytes).", K, params[0].memref.size);
+exit:
+	return res;
+}
+
+static TEE_Result get_hotp(uint32_t param_types, TEE_Param params[4])
+{
+	TEE_Result res = TEE_SUCCESS;
+	uint32_t hotp_val = 0;
```
Comment (jenswi-linaro) [inline hotp/ta/hotp_ta.c]: Initialization not needed.

File: hotp/host/main.c (hunk 11 above)
Comment (jbech-linaro) [reply hotp/host/main.c]: True, but in general it is considered good practice to initialize variables. From what I recall people generally agrees with that besides if it is a static variable (and it should be initialized to zero and nothing else).

File: hotp/ta/hotp_ta.c (hunk 13 above)
Comment (jbech-linaro) [reply hotp/ta/hotp_ta.c]: I prefer to keep it, since I think it makes it easier to grasp the overall code since you will see all the GP magic in 15 lines, which otherwise would have spread out over ~30 lines instead. 

> @jenswi-linaro Then there's also the issue with ret_val being evaluated three times, doesn't seem proper since it's really a function call with side effects.

Ah! True, that must be changed.

Comment (jbech-linaro) [reply hotp/ta/hotp_ta.c]: I'll remove the macro's so everyone is happy.

**Comment by jbech-linaro** [comment]
Pushed one more review commit fixing fixable checkpatch errors.

File: hotp/host/main.c (hunk 11 above)
Comment (jenswi-linaro) [reply hotp/host/main.c]: Needless initialization should be avoided since you otherwise start to wonder "what am I missing, why is this variable initialized?"
When it comes to zero initialize static variables, I thought the general consensus was the opposite!

Comment (jbech-linaro) [reply hotp/host/main.c]: > Needless initialization should be avoided since you otherwise start to wonder "what am I missing, why is this variable initialized?"

I disagree, but I will update to make everyone happy.

> When it comes to zero initialize static variables, I thought the general consensus was the opposite!

What? Are you saying that we are supposed to initialize static variables that should have the initial value as zero? If that is the case then I've learned something new today.

**Comment by jbech-linaro** [comment]
I think I've addressed all comment besides the bike-shedding comments.

File: hotp/host/main.c (hunk 11 above)
Comment (jenswi-linaro) [reply hotp/host/main.c]: Sorry, I misread you're previous comment.

Comment (jforissier) [reply hotp/host/main.c]: We're following the Linux kernel coding style so the rule is: no unnecessary initialization. While this choice certainly has downsides [1], it is expected that there is a greater benefit in not artificially hiding paths to the compiler where a variable is used before having received the proper value.

[1] https://www.cc.gatech.edu/grads/k/klu38/publications/ubi-ndss17.pdf

Comment (jbech-linaro) [reply hotp/host/main.c]: The last review commit removed it, so I guess we are good to go?

Comment (jforissier) [reply hotp/host/main.c]: Yes

**Comment by jenswi-linaro** [comment]
Who's talking about shedding bikes? I wouldn't have minded if a few empty unneded lines where shedded though.

**Comment by jbech-linaro** [comment]
> Who's talking about shedding bikes? I wouldn't have minded if a few empty unneded lines where shedded though.

Sorry, didn't mean to sound harsh, but that is fine and thanks for mentioning them. What I meant to say is that we all have different personal preferences and situations where it is not wrong, then we should just let it be like the author wants it to be. In this case I'm the author here and I want the empty lines because I like them. Fair enough?

**Comment by jenswi-linaro** [comment]
Since this is only in an examples git I'll let this slide.
`Acked-by: Jens Wiklander <jens.wiklander@linaro.org>`

**Comment by jbech-linaro** [comment]
Squashed, rebased and tag(s) applied (should be) ready for merge! Thanks!
